### PR TITLE
Misc changes 2

### PR DESCRIPTION
**Pull request by DanieleDiBenedetto**
- Avoid deriving CanonicalSerialize/CanonicalDeserialize for relevant structs and overriding default implementations with concrete ones to get the expected behaviour (e.g. deserialize_uncompressed calls the deserialize(compressed) if the function is not overridden, but that's not what we may want) (please review this carefully, we want to make sure that no (unexpected) default implementation is called);
- Making sure that `canonical_serialize_without_metadata()` will act like the previous ToBytes/FromBytes (preparing for a future PR in which we are going to remove them): meaning all data is saved in uncompressed form and no info regarding length or types of data structures are saved (please review this carefully too, as we currently use the to_bytes! in Fiat Shamir sponges);
- Add extra ci jobs for asm and target-feature;
- Added missing assertions on poly degrees;
- Refactored `index()` function in `circuit_specific_setup()`and `get_index_info()` functions;
- Refactored `IndexInfo.num_variables` in two new fields:  `num_witness` and `num_inputs`;
- Added new negative test case where proof fails in hard part;
- Fixed issue #12

### CODE REVIEW

File: src/ahp/indexer.rs [hunk 1]
```diff
@@ -20,8 +20,10 @@ use std::marker::PhantomData;
 #[derivative(Clone(bound = ""), Copy(bound = ""), Default(bound = ""), Debug(bound = ""), Eq(bound = ""), PartialEq(bound = ""))]
 #[derive(CanonicalSerialize, CanonicalDeserialize)]
 pub struct IndexInfo<F> {
-    /// The total number of variables in the constraint system.
-    pub num_variables: usize,
+    /// The total number of witnesses in the constraint system.
+    pub num_witness: usize,
```
Comment (UlrichHaboeck75) [inline src/ahp/indexer.rs:24]: maybe worth to mention: the number of witnesses includes constants.

Comment (UlrichHaboeck75) [reply src/ahp/indexer.rs:24]: aaah, sorry. Shouldn't be the constants to be treated as public inputs, always enforced to be equal to 1 in the verifier?

File: src/lib.rs [hunk 2]
```diff
@@ -90,17 +91,25 @@ impl<F: PrimeField, PC: PolynomialCommitment<F>, D: Digest> Marlin<F, PC, D> {
...
     }
 
-    /// Generate the index-specific (i.e., circuit-specific) prover and verifier
-    /// keys. This is a deterministic algorithm that anyone can rerun.
-    pub fn index<C: ConstraintSynthesizer<F>>(
-        committer_key: &PC::CommitterKey,
-        c: C,
-    ) -> Result<(ProverKey<F, PC>, VerifierKey<F, PC>), Error<PC::Error>> {
+    /// Wrapper around `AHPForR1CS::index`, for external access
+    pub fn get_index_info<C: ConstraintSynthesizer<F>>(c: C) -> Result<Index<F>, Error<PC::Error>> {
```
Comment (UlrichHaboeck75) [inline src/lib.rs:95]: A bit misleading name: ``get_index_info`` does re-run the indexer, hence it is not merely a lookup.

**Review by UlrichHaboeck75 — APPROVED** [review]

File: src/ahp/indexer.rs (hunk 1 above)
Comment (DanieleDiBenedetto) [reply src/ahp/indexer.rs:24]: I think constants are treated as witnesses, and enforced to be equal to the 1 "input" of the CS

File: src/lib.rs (hunk 2 above)
Comment (DanieleDiBenedetto) [reply src/lib.rs:95]: You're right but I don't want to confuse it with 'index'. Actually index function should be renamed according to me but I don't want to do it now cause I'll need to refactor in too many places

**Review by albertog78 — APPROVED** [review]

**Pull request merged**